core_lib/src/interface/editor.cpp

```diff
@@ -508,108 +502,193 @@ void Editor::resetAutoSaveCounter()
+
+            // Select and move any frames that may come into contact with the new position
+            currentLayer->selectBatchOfConnectedFrames(newPosition);
+            currentLayer->moveSelectedFrames(1);
         }
-        else
+
+        // It's a bug if the keyframe is nullpr at this point...
+        Q_ASSERT(it->second != nullptr);
+
+        if (currentLayer->type() == Layer::SOUND)
         {
-            g_clipboardBitmapImage = bitmapImage->copy();  // copy the whole image
+            SoundClip* key = static_cast<SoundClip*>(it->second)->clone();
+            currentLayer->addKeyFrame(newPosition, key);
+            sound()->loadSound(key, key->fileName());
+        } else {
+            // TODO: undo/redo implementation
+            KeyFrame* keyClone = it->second->clone();
+            currentLayer->addKeyFrame(newPosition, keyClone);
         }
-        clipboardBitmapOk = true;
-        if (g_clipboardBitmapImage.image() != nullptr)
-            QApplication::clipboard()->setImage(*g_clipboardBitmapImage.image());
-    }
-    if (layer->type() == Layer::VECTOR)
-    {
-        clipboardVectorOk = true;
-        VectorImage *vectorImage = static_cast<LayerVector*>(layer)->getLastVectorImageAtFrame(currentFrame(), 0);
-        if (vectorImage == nullptr) { return; }
-        g_clipboardVectorImage = *vectorImage;  // copy the image
+        currentLayer->deselectAll();
     }
 }
 
 void Editor::paste()
 {
-    Layer* layer = mObject->getLayer(layers()->currentLayerIndex());
-    if (layer != nullptr)
-    {
-        if (layer->type() == Layer::BITMAP && g_clipboardBitmapImage.image() != nullptr)
-        {
-            backup(tr("Paste"));
+    Layer* currentLayer = layers()->currentLayer();
 
-            BitmapImage tobePasted = g_clipboardBitmapImage.copy();
-            qDebug() << "to be pasted --->" << tobePasted.image()->size();
-            if (select()->somethingSelected())
-            {
-                QRectF selection = select()->mySelectionRect();
-                if (g_clipboardBitmapImage.width() <= selection.width() && g_clipboardBitmapImage.height() <= selection.height())
-                {
-                    tobePasted.moveTopLeft(selection.topLeft());
-                }
-                else
-                {
-                    tobePasted.transform(selection, true);
-                }
-            }
-            auto pLayerBitmap = static_cast<LayerBitmap*>(layer);
-            mScribbleArea->handleDrawingOnEmptyFrame();
-            BitmapImage *bitmapImage = pLayerBitmap->getLastBitmapImageAtFrame(currentFrame(), 0);
-            Q_CHECK_PTR(bitmapImage);
-            bitmapImage->paste(&tobePasted); // paste the clipboard
-        }
-        else if (layer->type() == Layer::VECTOR && clipboardVectorOk)
-        {
+    Q_ASSERT(currentLayer != nullptr);
+
+    if (clipboards()->getClipboardFrames().empty()) {
+
+        if (clipboards()->canPaste(currentLayer)) {
```

Comment:
Good point.. I must have missed something or made a mistake while reworking something here... it should be an early return if it's not possible to paste and otherwise backup if it's not a frame selection, since we can't undo a timeline frame selection.